File: 523.continuous-subarray-sum.cpp
/*
 * @lc app=leetcode id=523 lang=cpp
 *
 * [523] Continuous Subarray Sum
 */
#include <vector>
#include <unordered_set>
using namespace std;
// @lc code=start
class Solution
{
public:
    bool checkSubarraySum(vector<int> &nums, int k)
    {
        int width = nums.size();
        unordered_set<int> seen;
        seen.insert(nums[0] % k);
        int shift = 0;
        for (int i = 1; i != width; ++i)
        {
            shift += nums[i];
            shift %= k;
            if (seen.count((-shift + k) % k))
            {
                return true;
            }
            seen.insert((-shift + k + nums[i]) % k);
        }
        return false;
    }
};
// @lc code=end

File: 523.continuous-subarray-sum.java
import java.util.HashSet;
import java.util.Set;

/*
 * @lc app=leetcode id=523 lang=java
 *
 * [523] Continuous Subarray Sum
 */

// @lc code=start
class Solution {
    public boolean checkSubarraySum(int[] nums, int k) {
        Set<Integer> seen = new HashSet<>();
        seen.add(nums[0] % k);
        int shift = 0;
        for (int i = 1; i != nums.length; ++i) {
            shift += nums[i];
            shift %= k;
            if (seen.contains((-shift + k) % k)) {
                return true;
            }
            seen.add((-shift + k + nums[i]) % k);
        }
        return false;
    }
}
// @lc code=end
